pdal/GDALUtils.cpp

```diff
@@ -774,6 +777,107 @@ OGRGeometry *createFromGeoJson(const std::string& s, std::string& srs)
     return newGeom;
 }
 
+
+std::vector<OGRGeometry*> fetchOGRGeometries(const NL::json ogr)
+{
+    const NL::json& layer = ogr.at("layer");
+    const NL::json& datasource = ogr.at("datasource");
+    const NL::json options = ogr.at("options");
+
+
+    registerDrivers();
+
+
+    char** papszDriverOptions = nullptr;
+    if (ogr.count("drivers"))
+    {
+
+        const NL::json& dops = ogr.at("drivers");
+        std::vector<std::string> driverOptions = dops.get<std::vector<std::string>>();
+        for(const auto& s: driverOptions)
+            papszDriverOptions = CSLAddString(papszDriverOptions, s.c_str());
+    }
+    std::vector<const char*> openoptions{};
+    char** papszOpenOptions = nullptr;
+    if (ogr.count("openoptions"))
+    {
+
+        const NL::json& oops = ogr.at("openoptions");
+        std::vector<std::string> openOptions = oops.get<std::vector<std::string>>();
+        for(const auto& s: openOptions)
+            papszOpenOptions = CSLAddString(papszOpenOptions, s.c_str());
+    }
+
+    std::string dsString = datasource.get<std::string>();
+    unsigned int openFlags = GDAL_OF_READONLY | GDAL_OF_VECTOR | GDAL_OF_VERBOSE_ERROR;
+    GDALDataset* ds = (GDALDataset*) GDALOpenEx( dsString.c_str(),
+                                                 openFlags,
+                                                 papszDriverOptions, papszOpenOptions, NULL );
+    CSLDestroy(papszDriverOptions);
+    CSLDestroy(papszOpenOptions);
+    if (!ds)
+        throw pdal_error("Unable to read datasource in fetchOGRGeometries " + datasource.dump() );
+
+
+    OGRLayer* poLayer(nullptr);
+
+    std::string lyrString = layer.get<std::string>();
+    poLayer = ds->GetLayerByName( lyrString.c_str() );
+
+    if (!poLayer)
+        throw pdal_error("Unable to read layer in fetchOGRGeometries " + layer.dump() );
+
+    std::vector<OGRGeometry*> output;
+
+    OGRFeature *poFeature (nullptr);
+    OGRGeometry* filterGeometry(nullptr);
+
+    poLayer->ResetReading();
+    if (ogr.count("sql"))
+    {
+        std::string dialect("OGRSQL");
+        if (options.count("dialect"))
+            dialect = options.at("dialect").get<std::string>();
+
+        if (options.count("geometry"))
+        {
+            std::string wkt_or_json = options.at("geometry").get<std::string>();
+            bool isJson = (wkt_or_json.find("{") != wkt_or_json.npos) ||
+                          (wkt_or_json.find("}") != wkt_or_json.npos);
+
+            std::string srs;
+            if (isJson)
+            {
+                filterGeometry = createFromGeoJson(wkt_or_json, srs);
```

Comment:
Seems leaky for filterGeometry.  Maybe we should change things so that createFrom... creates pdal::Geometry().